index.bs

```diff
@@ -5638,6 +5661,183 @@ To <dfn>update the response</dfn> given |session|, |command| and |command parame
+
+  1. If |collector info|["userContexts"] [=map/contains=] |user context|'s [=user context id=], return true.
+
+  1. Otherwise, return false.
+
+1. Return true.
+
+</div>
+
+<div algorithm>
+To <dfn>update navigable network collector sizes</dfn> given |session| and |top-level navigable|:
+
+1. Let |match| be false.
+
+1. Let |resource max size| be 0.
+
+1. Let |total max size| be 0.
+
+1. For each |collector info| of |session|'s [=network collectors=]:
+
+  1. If [=match collector info=] with |top-level navigable| and |collector info|:
+
+    1. Set |match| to true.
+
+    1. If |collector info|["<code>resourceMaxSize</code>"] is greater than |resource max size|
+       set |resource max size| to |collector info|["<code>resourceMaxSize</code>"].
+
+    1. If |collector info|["<code>totalMaxSize</code>"] is greater than |total max size|
+       set |total max size| to |collector info|["<code>totalMaxSize</code>"].
+
+1. If |match| is true:
+
+  1. Let |sizes| be a new [=/map=] matching the <code>network.CollectorSizes</code> production,
+     with the <code>resourceMaxSize</code> field set to |resource max size|
+     and the <code>totalMaxSize</code> field set to |total max size|.
+
+  1. [=map/Set=] |session|'s [=navigable network collector sizes=][|top-level navigable|] to |sizes|.
+
+1. Otherwise:
+
+  1. [=map/Remove=] |session|'s [=navigable network collector sizes=][|top-level navigable|]
+
+</div>
+
+<div algorithm>
+To <dfn>maybe collect network data</dfn> given |session|, |request| and |response|:
+
+1. Let |navigable| be null.
+
+1. If |request|'s [=request/window=] is an [=environment settings object=]:
+
+  1. Let |environment settings| be |request|'s [=request/window=]
+
+  1. If there is a [=/navigable=] whose [=active window=] is |environment
+     settings|' [=environment settings object/global object=], set |navigable|
+     to be that navigable.
+
+1. If |navigable| is null, return.
+
+1. Let |top-level navigable| be |navigable|'s [=navigable/top-level traversable=].
+
+1. If |session|'s [=navigable network collector sizes=] does not [=list/contains|contain=]
+   |top-level navigable| return.
+
+1. Let |response to collect| be null.
+
+1. Let |response size| be |response|'s [=response/body=]'s [=body/length=].
+
+1. Let |sizes| be |session|'s [=navigable network collector sizes=][|top-level navigable|].
+
+1. If |response|’s [=response/body=] is not null and
+   if |sizes|' <code>maxResourceSize</code> is greater than |response size|:
+
+  1. [=Allocate size to record data=] with |session|, |navigable|, |response size| and
+     |sizes|' <code>maxTotalSize</code>.
+
+  1. Let |processBody| given |nullOrBytes| be this step:
+
+    1. If |nullOrBytes| is not null, set |response to collect| to [=serialize protocol bytes=] with |nullOrBytes|.
+
+  1. Let |processBodyError| be this step: Do nothing.
+
+  1. [=Fully read=] |response|’s [=response/body=] given |processBody| and |processBodyError|.
```

Comment:
I wonder if we need to clone the body first (in which case it probably has to happen early in the lifecycle, before anything else has read the body)? Otherwise I'm concerned that if some other process has already ready the response body (which seems likely to be the common case) then at this point we end up not reading any data (or, worse, we stop the data going to platform APIs).

Reply:
Yes, I think a stream can only be read once, so we need something that [tees](https://streams.spec.whatwg.org/#tee-a-readable-stream) the stream before we read it. [Cloning](https://fetch.spec.whatwg.org/#concept-body-clone) the body does this, but we need to be sure to do it before anything else has read from it.

Reply:
In that case, I think we need to move the response started trigger in https://whatpr.org/fetch/1540.html#http-network-fetch to be done after step 12 (stream creation) or 14 (body creation). Right now, we are emitting ResponseStarted at step 7.5. But it's done in a loop so I am not sure we can move it to a later step because we might miss some events.

Otherwise, if we want to decouple this from our events, Fetch can call BiDi to `maybe clone the response body` when the body is created? In any case it sounds like this will require some changes to the Fetch spec. We should agree on the design here and decide how we want to propose that to fetch (ie either new PR or reuse the existing one which adds the BiDi events)